src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -361,6 +361,444 @@ BaseFloat DecodableAmNnetSimpleParallel::LogLikelihood(int32 frame,
+    } else { 
+      first_subsampled_frame = past_last_subsampled_frame + 1;
+    }
+  }
+
+  int32 subsampling_factor = opts_.frame_subsampling_factor,
+        subsampled_frames_per_chunk = opts_.frames_per_chunk / subsampling_factor,
+        num_subsampled_frames = num_subsampled_frames_.find(utt_id)->second;
+  KALDI_ASSERT(num_subsampled_frames > 0);
+
+  int32 extra_left_context = opts_.extra_left_context,
+        extra_right_context = opts_.extra_right_context;
+
+  for(; first_subsampled_frame < num_subsampled_frames && !Ready();
+        first_subsampled_frame = last_subsampled_frame + 1) {
+    // Prepare first_subsampled_frame, last_subsampled_frame. They are the
+    // indexes of the output matrix.
+    int32 cur_num_subsampled_frames = 
+      std::min<int32>(subsampled_frames_per_chunk,
+                      num_subsampled_frames - first_subsampled_frame);
+    last_subsampled_frame = first_subsampled_frame + cur_num_subsampled_frames - 1;
+    // Prepare first_input_frame, last_input_frame 
+    int32 first_output_frame = first_subsampled_frame * subsampling_factor,
+          last_output_frame = last_subsampled_frame * subsampling_factor;
+    
+    if(first_output_frame == 0 && opts_.extra_left_context_initial >= 0) {
+      extra_left_context = opts_.extra_left_context_initial;
+    }
+    if (last_subsampled_frame == num_subsampled_frames - 1 &&
+        opts_.extra_right_context_final >= 0) {
+      extra_right_context = opts_.extra_right_context_final;
+    }
+
+    int32 left_context = nnet_left_context_ + extra_left_context;
+    int32 right_context = nnet_right_context_ + extra_right_context;
+    int32 first_input_frame = first_output_frame - left_context;
+    int32 last_input_frame = last_output_frame + right_context;
+    int32 left_zero_frames = std::max(opts_.extra_left_context,
+                                      opts_.extra_left_context_initial) -
+                             extra_left_context;
+      
+    // Update class private member
+    last_batch_info_ = std::make_tuple(utt_id,
+        first_input_frame, last_input_frame,
+        first_subsampled_frame, last_subsampled_frame, left_zero_frames);
+    batch_info_queue_.push(last_batch_info_);
+  }
+  // recursive invocation.
+  PrepareBatchInfo();
+}
+
+
+void BatchNnetComputer::DoNnetComputation() {
+  KALDI_ASSERT(!batch_info_queue_.empty() && 
+      "Please prepare the batch information fisrt!");
+  
+  std::queue<std::tuple<std::string, int32, int32, int32, int32, int32>> output_info_queue;
+
+  int32 subsampling_factor = opts_.frame_subsampling_factor;
+
+  Matrix<BaseFloat> tot_input(num_batch_input_rows_ * minibatch_size_,
+                              nnet_.InputDim("input"), kSetZero);
+  Matrix<BaseFloat> tot_ivector;
+  int32 ivector_dim = nnet_.InputDim("ivector");
+  if (ivector_dim > 0) {
+    tot_ivector.Resize(minibatch_size_, ivector_dim, kSetZero);
+  }
+
+  // 1.Prepare data
+  for(int32 n=0; !batch_info_queue_.empty(); n++) {
```

Comment:
wrong indentation n = 0

Reply:
Thanks

Reply:
and need space after for.   See if misc/maintenance/cpplint.py still works.

Reply:
Ok, I will use cpplint to check all the code.